### PR TITLE
First steps at implementing a ADT for templates

### PR DESCRIPTION
**Pull request by mverardo**
First steps towards #28.

This PR:
- Creates a TemplateFactory, which contains BuildTemplate[] (a syntactic sugar to create the corresponding template Association).
- Introduces a new subpackage "Expansion", with new implementations of the old expansion functions (code much better organized, simpler and more elegant).
- Creates a new ExpandTemplate function, which dispatches to the tight expansion function of a given template.
- Adds an overload of RuleTemplateVars that works on the new ADT.

Next steps towards #28 would be:
- Adapt all generator functions to build the template ADT instead of a simple list.
- Create overloads on Basic's functions to work with the ADT instead of a simple list, and drop the old definitions once they are no longer needed.
- Figure out how to decide which expansion function should be the result of a intersection.
- Figure out how to expand constrained variables using this new implementation.

### CODE REVIEW

**Comment by mverardo** [comment]
@zorandir, could you take a look? :smile:

File: TemplateGeneration/TemplateFactory.m [hunk 1]
```diff
@@ -0,0 +1,26 @@
+(* ::Package:: *)
+
+BeginPackage["CATemplates`TemplateGeneration`TemplateFactory`"];
+
+
+BuildTemplate::usage=
+"BuildTemplate[rawList_List, k_Integer, r_Real, expansion_Function]
```
Comment (zorandir) [inline TemplateGeneration/TemplateFactory.m]: The order of the arguments is wrong

File: Basic.wl [hunk 2]
```diff
@@ -3,6 +3,10 @@
 BeginPackage["CATemplates`Basic`"];
 
 
+Partial::usage = "Partial[f_, args__] := partially applies arguments args to function f."
```
Comment (zorandir) [inline Basic.wl]: Missed the semi-comma in the end of line

File: Basic.wl [hunk 3]
```diff
@@ -46,27 +50,24 @@ PreservesIndexVariableDualityQ::usage = "PreservesIndexVariableDualityQ[template
 
 ConstantsToVariables::usage = "ConstantsToVariables[replacementRules_]: Receives a list of replacement rules, and converts any symbol of the type C[i_Integer] into its corresponding template variable, preserving the index-variable duality."
 
-
 Begin["`Private`"];
 
+ClearAll@Partial;
```
Comment (zorandir) [inline Basic.wl]: This line remove the "usage" atribute from Partial method:

> In: ?Partial
> Out: CATemplates`Basic`Partial
>   Attributes[Partial]={HoldAll}
> Partial[CATemplates`Basic`Private`f_,CATemplates`Basic`Private`as__]:=Function[Null,CATemplates`Basic`Private`f[CATemplates`Basic`Private`as,##1],HoldAll]

**Comment by zorandir** [comment]
This pull request looks ok for me.

**Comment by mverardo** [comment]
@PPBO could you also take a look?

**Pull request merged**